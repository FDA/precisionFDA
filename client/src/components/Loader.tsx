/* eslint-disable react/require-default-props */
import React from 'react'
import styled, { css } from 'styled-components'
import { colors } from '../styles/theme'
import { Svg } from './icons/Svg'

const LoaderWrapper = styled.div<{shouldDisplayInline?: boolean}>`
  display: flex;
  justify-content: center;
  margin-top: 16px;
  ${({ shouldDisplayInline }) => shouldDisplayInline && css`
    display: inline;
  `}
`
type Props = {
  height?: number
  shouldDisplayInline?: boolean
}

export const Loader = ({ height = 16, shouldDisplayInline }: Props) => (
  <LoaderWrapper shouldDisplayInline={shouldDisplayInline}>
    <Svg height={height} width={60}>
      <circle fill={colors.darkBlue} stroke="none" cx="6" cy={height / 2} r="6">
        <animate
          attributeName="opacity"
          dur="1s"
          values="0;1;0"
          repeatCount="indefinite"
          begin="0.1"
        />
      </circle>
      <circle fill={colors.darkBlue} stroke="none" cx="26" cy={height / 2} r="6">
        <animate
          attributeName="opacity"
          dur="1s"
          values="0;1;0"
          repeatCount="indefinite"
          begin="0.2"
        />
      </circle>
      <circle fill={colors.darkBlue} stroke="none" cx="46" cy={height / 2} r="6">
        <animate
          attributeName="opacity"
          dur="1s"
          values="0;1;0"
          repeatCount="indefinite"
          begin="0.3"
        />
      </circle>
    </Svg>
  </LoaderWrapper>
)
