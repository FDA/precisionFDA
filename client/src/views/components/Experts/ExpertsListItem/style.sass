@import "../../../../styles/common.sass"
@import "../../../../styles/variables.sass"


h1
  @extend .pfda-section-heading
  margin: 4px 0 4px 0

.expert-name
  font-weight: bold
  font-size: $pfda-font-size-body
  padding-right: 12px

  a
    color: $color-text-black

.expert-date
  font-size: 13px
  color: $color-text-medium-grey

.experts-list-item
  display: flex
  flex-flow: row nowrap
  list-style: none
  margin-bottom: $padding-main-content-vertical

  &__left-column
    width: $sizing-smaller-column-width
    flex: 0 0 $sizing-smaller-column-width
    align-items: flex-start
    padding: 0

  &__right-column
    flex-grow: 1
    align-items: flex-start
    padding-left: 0
    padding-right: $padding-main-content-horizontal

  .expert-image
    width: $sizing-smaller-column-width
    height: $sizing-smaller-column-width
    object-fit: cover
    overflow: hidden

  .experts-item-content
    flex-grow: 1
    vertical-align: top
    padding: 0 0 0 $padding-main-content-horizontal

    h1
      @extend .pfda-section-heading
      margin-bottom: 6px

    p
      font-size: $pfda-font-size-body
      font-weight: 400
      color: $color-text-medium-grey
      margin-top: 5px

  &__buttons
    button
      margin-right: $padding-button-spacing * 1.5

    a
      margin-right: $padding-button-spacing * 1.5


  @media (min-width: 768px)
    flex-flow: row nowrap


.experts-list-item-small
  display: flex
  margin-bottom: $padding-main-content-vertical / 2

  &__left-column
    width: 60x
    flex: 0 0 60px
    align-items: flex-start
    padding: 0

  &__right-column
    flex-shrink: 1
    align-items: center
    padding: 0 0 0 12px
    margin: auto 0
    font-size: 12px

  .expert-image
    width: $sizing-icon-small
    height: $sizing-icon-small
    margin: auto
    object-fit: cover
    overflow: hidden

  .expert-name
    font-size: 13px
    margin-bottom: 2px

  .expert-blog-title
    margin: 3px 0px

  .expert-date
    font-size: 12px

  .answers-comments-container
    display: flex
    justify-content: space-between

    div
      width: 50%

    .answers-comments
      color: $color-text-medium-grey
      margin-left: 3px
      margin-right: 6px

.btn-borderless
  border: none
