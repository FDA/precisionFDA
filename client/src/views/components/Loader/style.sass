.pfda-loader
  font-size: 36px
