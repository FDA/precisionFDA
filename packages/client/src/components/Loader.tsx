/* eslint-disable react/require-default-props */
import React from 'react'
import styled from 'styled-components'
import { Svg } from './icons/Svg'

export const LWrap = styled.div`
  align-self: stretch;
  height: 100%;
`

export const LoaderMargin = styled.div`
  margin: 24px;
`

const LoaderWrapper = styled.div`
  display: flex;
  justify-content: center;
  &.inline {
    display: inline;
  }
  &.pageloader {
    padding-top: 16px;
  }
`
type Props = {
  height?: number
  className?: string
}

export const Loader = ({ height = 16, ...rest }: Props) => {
  // Radius of circles; Max 6px, but smaller in small loaders
  const radius = height >= 12 ? 6 : height / 2

  // Width; max 60px
  const width = radius * 10

  // Horizontal position of circles
  const cx1 = 6 * width / 60
  const cx2 = 26 * width / 60
  const cx3 = 46 * width / 60

  return (
    <LoaderWrapper {...rest}>
      <Svg height={height} width={width}>
        <circle fill="var(--base)" stroke="none" cx={cx1} cy={height / 2} r={radius}>
          <animate
            attributeName="opacity"
            dur="1s"
            values="0;1;0"
            repeatCount="indefinite"
            begin="0.1"
          />
        </circle>
        <circle fill="var(--base)" stroke="none" cx={cx2} cy={height / 2} r={radius}>
          <animate
            attributeName="opacity"
            dur="1s"
            values="0;1;0"
            repeatCount="indefinite"
            begin="0.2"
          />
        </circle>
        <circle fill="var(--base)" stroke="none" cx={cx3} cy={height / 2} r={radius}>
          <animate
            attributeName="opacity"
            dur="1s"
            values="0;1;0"
            repeatCount="indefinite"
            begin="0.3"
          />
        </circle>
      </Svg>
    </LoaderWrapper>
  )
}
