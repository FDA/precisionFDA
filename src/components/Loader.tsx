/* eslint-disable react/require-default-props */
import React from 'react'
import styled, { css } from 'styled-components'
import { colors } from '../styles/theme'
import { Svg } from './icons/Svg'

const LoaderWrapper = styled.div<{displayInline?: boolean}>`
  display: flex;
  justify-content: center;
  padding-top: 16px;
  ${({ displayInline }) => displayInline && css`
    display: inline;
  `}
`
type Props = {
  height?: number
  displayInline?: boolean
}

export const Loader = ({ height = 16, displayInline }: Props) => (
  <LoaderWrapper displayInline={displayInline}>
    <Svg height={height} width={60}>
      <circle fill={colors.darkBlue} stroke="none" cx="6" cy={height / 2} r="6">
        <animate
          attributeName="opacity"
          dur="1s"
          values="0;1;0"
          repeatCount="indefinite"
          begin="0.1"
        />
      </circle>
      <circle fill={colors.darkBlue} stroke="none" cx="26" cy={height / 2} r="6">
        <animate
          attributeName="opacity"
          dur="1s"
          values="0;1;0"
          repeatCount="indefinite"
          begin="0.2"
        />
      </circle>
      <circle fill={colors.darkBlue} stroke="none" cx="46" cy={height / 2} r="6">
        <animate
          attributeName="opacity"
          dur="1s"
          values="0;1;0"
          repeatCount="indefinite"
          begin="0.3"
        />
      </circle>
    </Svg>
  </LoaderWrapper>
)
