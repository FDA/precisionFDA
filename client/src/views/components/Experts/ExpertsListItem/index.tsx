import React, { Component } from 'react'
import { Link } from 'react-router-dom'
import classNames from 'classnames/bind'

import { IExpert } from '../../../../types/expert'
import Button from '../../Button'
import { format } from 'date-fns'
import { getSpacesIcon } from '../../../../helpers/spaces'
import Icon from '../../Icon'
import './style.sass'

type ExpertsListItemProps = {
  expert: IExpert
  userCanEdit: boolean
}

class ExpertsListItem extends Component<ExpertsListItemProps> {}

// ExpertsListItemBlogEntry - Used in /experts as the main list
class ExpertsListItemBlogEntry extends ExpertsListItem {
  render() {
    const expert = this.props.expert
    const userCanEdit = this.props.userCanEdit
    const classes = classNames('experts-list-item')
    return (
      <div className={classes}>
        <div className="experts-list-item__left-column">
          <img
            className="expert-image img-circle"
            src={expert.image}
            alt={`Profile icon for ${expert.title}'s blog`}
          />
        </div>
        <div className="experts-list-item__right-column experts-item-content">
          <h1>{expert.blogTitle}</h1>
          <div style={{ marginTop: '4px' }}>
            <span className="expert-name">{expert.title}</span>
            <span className="expert-date">
              {format(expert.createdAt, 'MMM dd, yyyy')}
            </span>
          </div>
          <p>{expert.blogPreview}</p>
          <div
            className="experts-list-item__buttons pull-left"
            style={{ marginTop: '6px' }}
          >
            <Button
              className=""
              size="md"
              type="default"
              onClick={() => window.location.assign(`/experts/${expert.id}/qa`)}
            >
              Expert Q&amp;A
            </Button>
            <Link to={`/experts/${expert.id}`}>
              <Icon icon={getSpacesIcon('experts')} fw />
              <span aria-label={`Click to view more information about ${expert.title}`}>About This Expert</span>
            </Link>
            <Link to={`/experts/${expert.id}/blog`}>
              <span aria-label={`Click to read ${expert.title}'s blog post`}>Read Expert Blog Post &#x2197;</span>
            </Link>
          </div>
          {userCanEdit && (
            <div className="btn-group pull-right" style={{ marginTop: '4px' }}>
              <a
                className="btn btn-default"
                href={`/experts/${expert.id}/edit`}
                aria-label={`Click here to Edit the Expert ${expert.id}`}
              >
                <span className="fa fa-pencil"></span> Edit Expert
              </a>
              <a
                className="btn btn-default"
                href={`/experts/${expert.id}/dashboard`}
                aria-label={`Click here to View ${expert.id}'s Dashboard`}
              >
                <span className="fa fa-dashboard fa-fw"></span> Dashboard
              </a>
            </div>
          )}
        </div>
      </div>
    )
  }
}

// ExpertsListItemBlogEntrySmall - Used in landing page on the side bar
class ExpertsListItemBlogEntrySmall extends ExpertsListItem {
  render() {
    const expert = this.props.expert
    const classes = classNames('experts-list-item-small')
    return (
      <div className={classes}>
        <div className="experts-list-item-small__left-column">
          <img
            className="expert-image img-circle"
            src={expert.image}
            alt={`Profile icon for ${expert.title}'s blog`}
          />
        </div>
        <div className="experts-list-item-small__right-column experts-item-content">
          <div className="expert-name">{expert.title}</div>
          <div className="expert-blog-title">{expert.blogTitle}</div>
          <div className="expert-date">
            {format(expert.createdAt, 'MMM dd, yyyy')}
          </div>
        </div>
      </div>
    )
  }
}

// ExpertsListItemQuestionsAndAnswers - Used in /experts in the side bar
class ExpertsListItemQuestionsAndAnswers extends ExpertsListItem {
  render() {
    const expert = this.props.expert
    const classes = classNames('experts-list-item-small')
    const showQuestionsCommentsCount =
      expert.totalAnswerCount > 0 || expert.totalCommentCount > 0
    return (
      <div className={classes}>
        <div className="experts-list-item-small__left-column">
          <img
            className="expert-image img-circle"
            src={expert.image}
            alt={`Profile icon for ${expert.title}'s blog`}
          />
        </div>
        <div className="experts-list-item-small__right-column experts-item-content">
          <div className="expert-name">
            <Link to={`/experts/${expert.id}`}>
              <span>{expert.title}</span>
            </Link>
          </div>
          {showQuestionsCommentsCount && (
            <div className="answers-comments-container">
              <div>
                <span>{expert.totalAnswerCount}</span>&nbsp;
                <span className="answers-comments">Answers</span>
              </div>
              <div>
                <span>{expert.totalCommentCount}</span>&nbsp;
                <span className="answers-comments">Comments</span>
              </div>
            </div>
          )}
        </div>
      </div>
    )
  }
}

export {
  ExpertsListItem,
  ExpertsListItemBlogEntry,
  ExpertsListItemBlogEntrySmall,
  ExpertsListItemQuestionsAndAnswers,
}

export default ExpertsListItem
