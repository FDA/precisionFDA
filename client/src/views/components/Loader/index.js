import React from 'react'

import Icon from '../Icon'
import './style.sass'


const Loader = () => (
  <Icon icon="fa fa-spinner fa-spin" cssClasses="pfda-loader" />
)

export default Loader
