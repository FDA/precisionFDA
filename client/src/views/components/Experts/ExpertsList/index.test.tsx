import React from 'react'
import { BrowserRouter } from 'react-router-dom'
import { shallow, mount } from 'enzyme'

import Loader from '../../Loader'
import { ExpertsList } from '.'
import { ExpertsListItemBlogEntry } from '../ExpertsListItem'


const getMockExperts = () => {
  let mockExperts = []
  const dateNow = new Date()

  for (let i=0; i<10; i++) {
    mockExperts.push({
      id: i,
      user_id: i,
      state: 'open',
      scope: 'public',
      image: '',
      about: '',
      title: 'Expert '+i,
      blogTitle: 'Blog Title '+i,
      blogPreview: 'Blog Preview '+i,
      blog: 'Blog '+i,
      createdAt: dateNow,
      updatedAt: dateNow,
      totalAnswerCount: 10,
      totalCommentCount: 11,
    })
  }
  return mockExperts
}


describe('ExpertsList test', () => {
  it('should render', () => {
    const wrapper = shallow(<ExpertsList />)

    expect(wrapper).toMatchSnapshot()
  })

  it('should show loader when isFetching and not show ExpertsList', () => {
    const wrapper = mount(<ExpertsList isFetching={true} listItemComponent={ExpertsListItemBlogEntry} />)

    // console.log(wrapper.debug())
    expect(wrapper.find(Loader)).toHaveLength(1)
    expect(wrapper.find('.experts-list')).toHaveLength(0)
    expect(wrapper.find('.experts-list-item')).toHaveLength(0)
  })

  it('should not show loader when not fetching and show ExpertsList with no rows', () => {
    const wrapper = mount(<ExpertsList experts={[]} isFetching={false} listItemComponent={ExpertsListItemBlogEntry} />)
    // console.log(wrapper.debug())

    expect(wrapper.find(Loader)).toHaveLength(0)
    expect(wrapper.find('div.text-center').text()).toEqual('No experts found.')
    expect(wrapper.find({ text: 'No experts found.'})).toHaveLength(0)
    expect(wrapper.find('ul')).toHaveLength(0)
    expect(wrapper.find('li')).toHaveLength(0)
  })

  it('should render 10 rows', () => {
    const mockExperts = getMockExperts()
    const mockPagination = {
      currentPage: 1,
      totalPages: 2,
      nextPage: 1,
      prevPage: 2,
      totalCount: 20,
    }
    const wrapper = mount (
      <BrowserRouter>
        <ExpertsList experts={mockExperts} pagination={mockPagination} isFetching={false} listItemComponent={ExpertsListItemBlogEntry} />
      </BrowserRouter>
    )
    // console.log(wrapper.debug())

    // Test items
    expect(wrapper.find('ul')).toHaveLength(1)
    expect(wrapper.find('.experts-list')).toHaveLength(1)
    expect(wrapper.find('.experts-list-item')).toHaveLength(10)

    // Test item props
    const items = wrapper.find(ExpertsListItemBlogEntry)
    expect(items).toHaveLength(10)
    expect(items.at(9).props().expert.id).toEqual(9)
    expect(items.at(8).props().expert.title).toEqual('Expert 8')
  })
})
