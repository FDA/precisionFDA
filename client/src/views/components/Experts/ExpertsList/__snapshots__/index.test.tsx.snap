// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExpertsList test should render 1`] = `
<div
  className="text-center"
>
  No experts found.
</div>
`;
